src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs

```diff
@@ -0,0 +1,269 @@
+
+            // This matches the definition locations to syntax references, ignores metadata locations
+            IEnumerable<Location> partialLocations = from currentSyntaxNode in syntaxNodes
+                                                     from sourceLocation in symbol.Locations
+                                                     where !sourceLocation.IsInMetadata
+                                                           && currentSyntaxNode.SyntaxTree.Equals(sourceLocation.SourceTree)
+                                                           && currentSyntaxNode.Span.Contains(sourceLocation.SourceSpan)
+                                                     select sourceLocation;
+
+            return partialLocations;
+        }
+
+        private static IEnumerable<Location> ExtractLocations(Tuple<SyntaxNode, IEnumerable<Tuple<ReferencedSymbol, bool>>> referenceInfo, CancellationToken cancellationToken)
+        {
+            SyntaxNode definitionSyntaxNode = referenceInfo.Item1;
+            IEnumerable<Tuple<ReferencedSymbol, bool>> referencingSymbols = referenceInfo.Item2;
+
+            // Take reference locations from all definitions
+            IEnumerable<Location> referenceLocations = from referencedSymbol in referencingSymbols
+                                                       from location in referencedSymbol.Item1.Locations
+                                                       select location.Location;
+
+            // Exclude the definition we queried for - base references should be 0
+            IEnumerable<ISymbol> definitions = from referencedSymbol in referencingSymbols
+                                               where !referencedSymbol.Item2    // Item2 indicates if this was the queried symbol definition
+                                               select referencedSymbol.Item1.Definition;
+
+            IEnumerable<Location> definitionLocations = definitions.SelectMany(def => def.Locations);
+
+            // Partial types can have more than one declaring syntax references.
+            // Add remote locations for all the syntax references except the queried syntax node.
+            // To query for the partial locations, filter definition locations that occur in source whose span is part of
+            // span of any syntax node from Definition.DeclaringSyntaxReferences except for the queried syntax node.
+            IEnumerable<Location> additionalSyntaxLocations = from referencedSymbol in referencingSymbols
+                                                              let symbolReference = referencedSymbol.Item1
+                                                              let isQueriedDefinition = referencedSymbol.Item2
+                                                              where isQueriedDefinition
+                                                              from partialLocation in GetPartialLocations(symbolReference.Definition, definitionSyntaxNode, cancellationToken)
```

Comment:
feels like doing partial location search multiple times. not sure how, put I think I saw multiple call to GetParticalLocations.

also, finding Locations from symbol seems done multiple times.